// @cosmograph/ui/utils.js
import { format as D } from "d3-format";
const E = (t) => typeof t == "number", I = (t) => typeof t == "number" && Number.isFinite(t), T = (t) => typeof t == "function", $ = (t) => t === void 0, P = (t) => t == null, _ = (t) => typeof t == "string", F = (t) => Array.isArray(t), O = (t) => t instanceof Object, m = (t) => t.constructor.name !== "Function" && t.constructor.name !== "Object", w = (t) => O(t) && !F(t) && !T(t) && !m(t), x = (t) => t instanceof Date && !isNaN(t.getTime()), p = (t, e = /* @__PURE__ */ new Map()) => {
  if (typeof t != "object" || t === null)
    return t;
  if (t instanceof Date)
    return new Date(t.getTime());
  if (t instanceof Array) {
    const r = [];
    e.set(t, r);
    for (const n of t)
      r.push(e.has(n) ? e.get(n) : p(n, e));
    return t;
  }
  if (m(t))
    return t;
  if (t instanceof Object) {
    const r = {};
    e.set(t, r);
    const n = t;
    return Object.keys(t).reduce((s, f) => (s[f] = e.has(n[f]) ? e.get(n[f]) : p(n[f], e), s), r), r;
  }
  return t;
}, N = (t, e, r = /* @__PURE__ */ new Map()) => {
  const n = m(t) ? t : p(t);
  return t === e ? t : r.has(e) ? r.get(e) : (r.set(e, n), Object.keys(e).forEach((s) => {
    w(t[s]) && w(e[s]) ? n[s] = N(t[s], e[s], r) : m(e) ? n[s] = e : n[s] = p(e[s]);
  }), n);
}, L = (t, e, r) => t >= +e && +t <= +r, R = (t) => {
  const e = getComputedStyle(t);
  let r = t.clientWidth, n = t.clientHeight;
  return n -= parseFloat(e.paddingTop) + parseFloat(e.paddingBottom), r -= parseFloat(e.paddingLeft) + parseFloat(e.paddingRight), { height: n, width: r };
}, A = (t, e, r = /* @__PURE__ */ new Set()) => {
  if (Array.isArray(t)) {
    if (!Array.isArray(e) || t.length !== e.length)
      return !1;
    if (r.has(t))
      return !0;
    r.add(t);
    for (let n = 0; n < t.length; n++)
      if (!A(t[n], e[n], r))
        return !1;
    return !0;
  }
  if (t instanceof Date && e instanceof Date)
    return t.getTime() === e.getTime();
  if (typeof t == "object" && t !== null && e !== null) {
    if (typeof e != "object")
      return !1;
    if (t === e)
      return !0;
    if (Object.keys(t).length !== Object.keys(e).length)
      return !1;
    if (r.has(t))
      return !0;
    r.add(t);
    for (const s in t)
      if (!A(t[s], e[s], r))
        return !1;
    return !0;
  }
  return t === e;
}, V = (t, e = "", r, n) => {
  if (!t)
    return "0";
  let s, f;
  r ? (s = t, f = r) : Math.abs(t) >= 99e10 ? (s = t / 1e12, f = "T") : Math.abs(t) >= 99e7 ? (s = t / 1e9, f = "B") : Math.abs(t) >= 99e5 ? (s = t / 1e6, f = "M") : Math.abs(t) >= 9900 ? (s = t / 1e3, f = "K") : (s = t, f = "");
  const l = Math.abs(s);
  n === void 0 && (Number.isInteger(l) ? n = 0 : l < 0.1 ? n = 3 : l < 1 ? n = 2 : l < 10 ? n = 1 : n = 0);
  let a = l > 0 && l < 0.5 / Math.pow(10, n);
  if (a) {
    let o = n;
    for (; a && o < 8; )
      o++, a = l < 0.5 / Math.pow(10, o);
    if (!a)
      n = o;
    else
      return `~${e ?? ""}0`;
  }
  let i = t < 0 ? "-" : "";
  return e && (i += e), `${i}${D(`.${n}~f`)(l)}${f}`;
};
function S(t, e, r = e, n = []) {
  if (!t.length || !e.length || !r.length)
    return new Array(e.length).fill(0);
  const s = new Array(e.length).fill(0), f = /* @__PURE__ */ new Map();
  for (const [i, c] of r) {
    f.has(i) || f.set(i, /* @__PURE__ */ new Map());
    for (let o = 0; o < e.length; o++) {
      const [u, g] = e[o];
      if (i === u && c === g) {
        f.get(i).set(c, o);
        break;
      }
    }
  }
  const l = n.length > 0 ? n[n.length - 1] : e[e.length - 1][1], a = [...r].sort((i, c) => i[0] - c[0]);
  for (const i of t) {
    let c = 0, o = a.length - 1;
    for (; c <= o; ) {
      const u = Math.floor((c + o) / 2), [g, h] = a[u], M = h === l;
      if (i >= g && (M ? i <= h : i < h)) {
        const d = f.get(g);
        if (d) {
          const y = d.get(h);
          y !== void 0 && s[y]++;
        }
        break;
      } else i < g ? o = u - 1 : c = u + 1;
    }
  }
  return s;
}
function q(t, e, r) {
  if (!t)
    return [];
  const [n, s] = t, f = Math.abs(+s - +n), l = Number.isFinite(f) && e > 1 ? f / (e - 1) : 1, a = r ?? l;
  if (a === 0)
    return [];
  const i = Array.from({ length: e }, (c, o) => +n + o * a);
  return i[e - 1] = +s, i;
}
function H(t, e, r, n = !1, s = !0) {
  const f = Array.from({ length: e.length - 1 }, (i, c) => [e[c], e[c + 1]]), l = r && n ? f.filter(([i, c]) => {
    const [o, u] = r;
    return i < +o ? !1 : c >= +u ? c > +o && i < +u : !0;
  }) : f, a = S(t, f, l, e);
  return f.map(([i, c], o) => ({
    rangeStart: s ? i : new Date(i),
    rangeEnd: s ? c : new Date(c),
    count: a[o] ?? 0
  }));
}
export {
  p as cloneDeep,
  V as defaultFormatter,
  q as generateBandIntervals,
  R as getInnerDimensions,
  m as isAClassInstance,
  F as isArray,
  L as isBetween,
  A as isEqual,
  I as isFiniteNumber,
  T as isFunction,
  P as isNil,
  E as isNumber,
  O as isObject,
  w as isPlainObject,
  _ as isString,
  $ as isUndefined,
  x as isValidDate,
  H as mapDataToBars,
  S as mapDataToIntervals,
  N as merge
};
